python/ray/autoscaler/_private/aws/config.py

```diff
@@ -959,6 +963,78 @@ def _configure_node_cfg_from_launch_template(
     return node_cfg
 
 
+def _configure_cluster_event_notifications(
+        config: Dict[str, Any]) -> Dict[str, Any]:
+    """If `cluster_notifications` is provided, an SNS callback handler will be added to
+    cluster creation events. The callback handler will send the event ID and publish
+    a notification to the SNS topic specified at `cluster_notifications.sns_topic_arn`.
+    The SNS topic must be in the same AWS account associated to the autoscaler cluster role.
+
+    Message parameters can be specified in `cluster_notifications.parameters`, which
+    will be included into the payload of each SNS notification.
+
+    Args:
+        config (Dict[str, Any]): config to bootstrap
+
+    Returns:
+        config (Dict[str, Any]): The input config with cluster event notification
+        data merged into the node config of all available node types. If no
+        cluster event notification data is found, then the config is returned
+        unchanged.
+    """
+    # create a copy of the input config to modify
+    config = copy.deepcopy(config)
+    provider = config['provider']
+
+    # TODO: Need a way to generalize this so that users can plug in their own custom callback logic
+    #  or have a dynamic config
+
+    if provider and provider.get('cluster_notifications'):
+        cluster_notifications_map = provider['cluster_notifications']
+        parameters = cluster_notifications_map['parameters']
+        sns_topic_arn = cluster_notifications_map['sns_topic_arn']
```

Comment:
Let's change this to just be an ARN that we then instantiate the appropriate handler for SNS, Lambda, API Gateway, etc. I think we can start with just our existing SNS handler though (converted to a class).

Reply:
Alternatively, to make it more generic than just AWS, we could also just make this a URI.

Reply:
At that point, I think we can just move cluster_notifications above the provider level, and just make it a top-level config section that is common to all cloud providers. We can then open up an issue in Ray Core pointing back to these changes as a proposed general purpose event system for cluster launchers together w/ a high-level description of how we use it in prod, then follow up with a PR back to Ray Core.